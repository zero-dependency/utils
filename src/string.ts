/**
 * Generates an array of characters between the startChar and endChar, inclusive.
 *
 * @param {string} startChar
 * The first character of the array.
 *
 * @param {string} endChar
 * The last character of the array.
 *
 * @return {string[]}
 * An array of characters from startChar to endChar.
 *
 * @example
 * generateChars('a', 'd') // ['a', 'b', 'c', 'd']
 */
export function generateChars(startChar: string, endChar: string): string[] {
  const chars = []
  const j = endChar.charCodeAt(0)

  for (let i = startChar.charCodeAt(0); i <= j; ++i) {
    chars.push(String.fromCharCode(i))
  }

  return chars
}

/**
 * Capitalizes the first letter of a given string.
 *
 * @param {string} str
 * The string to capitalize.
 *
 * @return {string}
 * The capitalized string.
 *
 * @example
 * capitalize('hello') // 'Hello'
 */
export function capitalize(str: string): string {
  return str.charAt(0).toUpperCase() + str.slice(1)
}

/**
 * Returns a randomly generated token string.
 *
 * @return {string}
 * The token string generated by concatenating a random number and a base36 string after slicing off the first 2 characters.
 *
 * @example
 * randomToken() // '6c9qicypz1'
 */
export function randomToken(): string {
  return Math.random().toString(36).slice(2)
}

/**
 * Repeats every other character in the given string a specified number of times.
 *
 * @param {string} str
 * The input string to repeat characters from.
 *
 * @param {number} repeats
 * The number of times to repeat each character.
 *
 * @return {string}
 * The modified string with every other character repeated.
 *
 * @example
 * repeatEveryChars('abc', 2) // 'aabbcc'
 */
export function repeatEveryChars(str: string, repeats: number): string {
  let result = ''

  for (const char of str) {
    result += char.repeat(repeats)
  }

  return result
}

export interface DifferenceResult {
  index: number
  old?: string
  new?: string
}

export function getDifference(
  leftStr: string,
  rightStr: string
): DifferenceResult[] {
  const differenceResult: DifferenceResult[] = []
  if (leftStr === rightStr) return differenceResult

  for (
    let index = 0;
    index < Math.max(leftStr.length, rightStr.length);
    index++
  ) {
    if (leftStr[index] !== rightStr[index]) {
      differenceResult.push({
        index: index,
        old: leftStr[index],
        new: rightStr[index]
      })
    }
  }

  return differenceResult
}
