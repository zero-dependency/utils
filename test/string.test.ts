import { describe, expect, test } from 'vitest'
import { capitalize, generateChars, randomToken } from '../src/string.js'

describe('generateChars', () => {
  test('should be defined', () => {
    expect(generateChars).toBeDefined()
  })

  test('should return an array of characters', () => {
    expect(generateChars('a', 'c')).toEqual([
      'a',
      'b',
      'c'
    ])
    expect(generateChars('A', 'C')).toEqual([
      'A',
      'B',
      'C'
    ])
    expect(generateChars('0', '2')).toEqual([
      '0',
      '1',
      '2'
    ])
    expect(generateChars('a', 'a')).toEqual(['a'])
  })

  test('should return an empty array if `startChar` is greater than `endChar`', () => {
    expect(generateChars('c', 'a')).toEqual([])
    expect(generateChars('C', 'A')).toEqual([])
    expect(generateChars('2', '0')).toEqual([])
  })
})

describe('capitalize', () => {
  test('should be defined', () => {
    expect(capitalize).toBeDefined()
  })

  test('should convert a string to uppercase', () => {
    expect(capitalize('a')).toBe('A')
    expect(capitalize('A')).toBe('A')
    expect(capitalize('0')).toBe('0')
    expect(capitalize('aA0')).toBe('AA0')
  })

  test('should return an empty string if `str` is empty', () => {
    expect(capitalize('')).toBe('')
  })
})

describe('randomToken', () => {
  test('should be defined', () => {
    expect(randomToken).toBeDefined()
  })

  test('should return a random string', () => {
    for (let i = 0; i < 1000; i++) {
      expect(randomToken()).toMatch(/^[a-z0-9]{8,14}$/)
    }
  })
})
