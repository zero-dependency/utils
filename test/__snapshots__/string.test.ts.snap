// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getDifference > should return the difference between two strings 1`] = `
[
  {
    "index": 0,
    "new": "b",
    "old": "f",
  },
  {
    "index": 1,
    "new": "a",
    "old": "o",
  },
  {
    "index": 2,
    "new": "r",
    "old": undefined,
  },
]
`;

exports[`getDifference > should return the difference between two strings 2`] = `
[
  {
    "index": 3,
    "new": "o",
    "old": undefined,
  },
]
`;

exports[`getDifference > should return the difference between two strings 3`] = `
[
  {
    "index": 0,
    "new": "f",
    "old": undefined,
  },
  {
    "index": 1,
    "new": "o",
    "old": undefined,
  },
  {
    "index": 2,
    "new": "o",
    "old": undefined,
  },
]
`;

exports[`getDifference > should return the difference between two strings 4`] = `
[
  {
    "index": 0,
    "new": undefined,
    "old": "f",
  },
  {
    "index": 1,
    "new": undefined,
    "old": "o",
  },
  {
    "index": 2,
    "new": undefined,
    "old": "o",
  },
]
`;
